Assets/Mirror/Tests/WeaverTest.cs

```diff
@@ -0,0 +1,1012 @@
+        }
+
+        private void BuildAndWeaveTestAssembly(string baseName)
+        {
+            WeaverAssembler.OutputFile = baseName + ".dll";
+            WeaverAssembler.AddSourceFiles(new string[] { baseName + ".cs" });
+            WeaverAssembler.Build();
+
+            Assert.That(WeaverAssembler.CompilerErrors, Is.False);
+            if (m_weaverErrors.Count > 0)
+            {
+                Assert.That(m_weaverErrors[0], Does.StartWith("Mirror.Weaver error: "));
+            }
+        }
+        #endregion
+
+        #region Setup and Teardown
+        [OneTimeSetUp]
+        public void FixtureSetup()
+        {
+            // TextRenderingModule is only referenced to use TextMesh type to throw errors about types from another module
+            WeaverAssembler.AddReferencesByAssemblyName(new string[] { "UnityEngine.dll", "UnityEngine.CoreModule.dll", "UnityEngine.TextRenderingModule.dll", "Mirror.dll" });
+
+            CompilationFinishedHook.UnityLogEnabled = false;
+            CompilationFinishedHook.OnWeaverError += HandleWeaverError;
+            CompilationFinishedHook.OnWeaverWarning += HandleWeaverWarning;
+        }
+
+        [OneTimeTearDown]
+        public void FixtureCleanup()
+        {
+            CompilationFinishedHook.UnityLogEnabled = true;
+        }
+
+        [SetUp]
+        public void TestSetup()
+        {
+        }
+
+        [TearDown]
+        public void TestCleanup()
+        {
+            WeaverAssembler.DeleteOutputOnClear = true;
+            WeaverAssembler.Clear();
+
+            m_weaverWarnings.Clear();
+            m_weaverErrors.Clear();
+        }
+        #endregion
+
+        #region General tests
+        [Test] // -----------------------------------------------------------------------------------------
+        public void InvalidType()
+        {
+            BuildAndWeaveTestAssembly(TestContext.CurrentContext.Test.Name);
+
+            Assert.That(CompilationFinishedHook.WeaveFailed, Is.True);
+            Assert.That(m_weaverErrors.Count, Is.AtLeast(1));
+            Assert.That(m_weaverErrors[0], Does.Match("please make sure to use a valid type"));
+        }
+
+        [Test] // -----------------------------------------------------------------------------------------
+        public void RecursionCount()
+        {
+            BuildAndWeaveTestAssembly(TestContext.CurrentContext.Test.Name);
+
+            Assert.That(CompilationFinishedHook.WeaveFailed, Is.True);
+            Assert.That(m_weaverErrors.Count, Is.AtLeast(1));
+            Assert.That(m_weaverErrors[0], Does.Match("Check for self-referencing member variables"));
+        }
+
+        [Test] // -----------------------------------------------------------------------------------------
+        public void ClientGuardWrongClass()
+        {
+            BuildAndWeaveTestAssembly(TestContext.CurrentContext.Test.Name);
+
+            Assert.That(CompilationFinishedHook.WeaveFailed, Is.False);
```

Comment:
Double negative are hard to read.   Nothing you can do about it in this PR,   but I am thinking we should rename "WeaverFailed"  to "WeaverSucceded"  and flip it around everywhere in a future PR.

Reply:
Will be done in future PR then